common/include/common/blas2_state_counters.hpp

```diff
@@ -0,0 +1,267 @@
+  const double mem_readX = xlen;
+  const double mem_writeY = ylen;
+  const double mem_readY = (beta != scalar_t{0}) ? ylen : 0;
+  state.counters["bytes_processed"] =
+      (mem_readA + mem_readX + mem_writeY + mem_readY) * sizeof(scalar_t);
+  return;
+}
+
+template <Level2Op op, typename scalar_t, typename index_t>
+inline typename std::enable_if<op == Level2Op::ger>::type init_level_2_counters(
+    benchmark::State& state, const char* t_str, scalar_t beta = scalar_t{0},
+    index_t m = 0, index_t n = 0, index_t k = 0, index_t ku = 0,
+    index_t kl = 0) {
+  // Google-benchmark counters are double.
+  double m_d = static_cast<double>(m);
+  double n_d = static_cast<double>(n);
+  state.counters["m"] = m_d;
+  state.counters["n"] = n_d;
+  state.counters["n_fl_ops"] = 2 * m_d * n_d + std::min(n_d, m_d);
+  state.counters["bytes_processed"] =
+      (2 * m_d * n_d + m_d + n_d) * sizeof(scalar_t);
+  return;
+}
+
+template <Level2Op op, typename scalar_t, typename index_t>
+inline typename std::enable_if<op == Level2Op::sbmv>::type
+init_level_2_counters(benchmark::State& state, const char* t_str,
+                      scalar_t beta = scalar_t{0}, index_t m = 0, index_t n = 0,
+                      index_t k = 0, index_t ku = 0, index_t kl = 0) {
+  // Google-benchmark counters are double.
+  double k_d = static_cast<double>(k);
+  double n_d = static_cast<double>(n);
+  state.counters["k"] = k_d;
+  state.counters["n"] = n_d;
+
+  // Compute the number of A non-zero elements.
+  const double A_validVal = (n_d * (2.0 * k_d + 1.0)) - (k_d * (k_d + 1.0));
+
+  const double nflops_AtimesX = 2.0 * A_validVal;
+  const double nflops_timesAlpha = n_d;
+  const double nflops_addBetaY = (beta != scalar_t{0}) ? 2 * n_d : 0;
+  const double nflops_tot =
+      nflops_AtimesX + nflops_timesAlpha + nflops_addBetaY;
+  state.counters["n_fl_ops"] = nflops_tot;
+
+  const double mem_readA = A_validVal;
+  const double mem_readX = n_d;
+  const double mem_writeY = n_d;
+  const double mem_readY = (beta != scalar_t{0}) ? n_d : 0;
+  state.counters["bytes_processed"] =
+      (mem_readA + mem_readX + mem_writeY + mem_readY) * sizeof(scalar_t);
+  return;
+}
+
+template <Level2Op op, typename scalar_t, typename index_t>
+inline
+    typename std::enable_if<op == Level2Op::spmv || op == Level2Op::symv>::type
+    init_level_2_counters(benchmark::State& state, const char* t_str,
+                          scalar_t beta = scalar_t{0}, index_t m = 0,
+                          index_t n = 0, index_t k = 0, index_t ku = 0,
+                          index_t kl = 0) {
+  // Google-benchmark counters are double.
+  double n_d = static_cast<double>(n);
+  state.counters["n"] = n_d;
+  // Compute the number of A non-zero elements.
+  const double A_validVal = (n_d * (n_d + 1) / 2);
+
+  const double nflops_AtimesX = 2 * n_d * n_d;
+  const double nflops_timesAlpha = n_d;
+  const double nflops_addBetaY = (beta != scalar_t{0}) ? 2 * n_d : 0;
```

Comment:
Thanks for the refactoring @muhammad-tanvir-1211 ! 
I was just wondering why not add beta values to the state counters for benchmarks _(applies to other operators as well)_? especially that they're affecting computations & other state counters (flops & bytes processed). This might be relevant as two benchmarks for instance might happen to have the same nflops &/or bytes processed AND still have different beta values (one being zero) : in such case the benchmarking would compare two different operations (one with an addition beta!=0 and another without) but still will consider it as a unique operation given that beta value is absent from the record. What do you think?

Reply:
That's a good point. Changing the value of `beta` does have an impact on the overall performance as well since we may or may not need to read the output from global memory within the kernel. I'll add `beta` to the counters to reflect it's impact. Thanks for brining this up.

Reply:
Done.